File: day-03/part-1/skasch.py
import collections
from typing import Iterator
from tool.runners.python import SubmissionPy


class SkaschSubmission(SubmissionPy):

    def parse(self, s: str) -> Iterator[str]:
        for line in s.splitlines():
            if stripped_line := line.strip():
                yield stripped_line

    def run(self, s: str) -> int:
        """
        :param s: input in string format
        :return: solution flag
        """
        # Your code goes here
        counter = collections.Counter()
        n_lines = 0
        n_chars = None
        for line in self.parse(s):
            n_lines += 1
            if n_chars is None:
                n_chars = len(line)
            for idx, char in enumerate(line):
                if char == "1":
                    counter[n_chars - 1 - idx] += 1
        gamma = sum(1 << n for n, count in counter.items() if 2 * count > n_lines)
        epsilon = gamma ^ ((1 << n_chars) - 1)
        print(gamma, epsilon)
        return gamma * epsilon



def test_skasch():
    """
    Run `python -m pytest ./day-03/part-1/skasch.py` to test the submission.
    """
    assert (
        SkaschSubmission().run(
            """
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010
""".strip()
        )
        == 198
    )

File: day-03/part-1/skasch.rs
use std::env::args;
use std::time::Instant;

fn main() {
    let now = Instant::now();
    let output = run(&args().nth(1).expect("Please provide an input"));
    let elapsed = now.elapsed();
    println!("_duration:{}", elapsed.as_secs_f64() * 1000.);
    println!("{}", output);
}

fn parse(input: &str) -> Vec<&str> {
    input.lines().filter(|s| !s.is_empty()).collect()
}

fn get_gamma_epsilon(lines: Vec<&str>) -> (isize, isize) {
    let mut count_ones = vec![];
    let n_lines = lines.len();
    for line in lines {
        for (idx, char) in line.chars().enumerate() {
            match char {
                '1' => {
                    if count_ones.len() <= idx {
                        count_ones.push(1)
                    } else {
                        count_ones[idx] += 1
                    }
                }
                _ => (),
            }
        }
    }
    let size = count_ones.len();
    let mut res = 0;
    for value in count_ones {
        res = 2 * res + if 2 * value > n_lines { 1 } else { 0 };
    }
    (res, ((1 << size) - 1) ^ res)
}

fn run(input: &str) -> isize {
    // Your code goes here
    let (gamma, epsilon) = get_gamma_epsilon(parse(input));
    gamma * epsilon
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_test() {
        assert_eq!(
            run("00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010"),
            198
        )
    }
}
